File: 03-stream-processing/scala/src/main/scala/com/becomedataengineer/exercise2/StreamingJob.scala
package com.becomedataengineer.exercise2

import org.apache.spark.sql.types.{StringType, StructField, StructType}
import org.apache.spark.sql.{SparkSession, functions}

object StreamingJob {

  def main(args: Array[String]): Unit = {
    val sparkSession = SparkSession.builder().master("local[*]")
      .getOrCreate()
    import sparkSession.implicits._

    val dataFrame = sparkSession.readStream
      .format("kafka")
      .options(Map(
        "kafka.bootstrap.servers" -> "localhost:29092",
        "subscribe" -> "visits_raw",
        "startingOffsets" -> "EARLIEST"
      ))
      .load()

    val schemaForFiltering = StructType(Array(
      StructField("os", StringType),
      StructField("visitedPage", StringType)
    ))

    val recordsToWrite = dataFrame
      .select(
        $"key".cast("string"),
        functions.from_json($"value".cast("string"), schemaForFiltering).as("visit"),
        $"value".cast("string")
      )
      .filter("visit.os IN ('macOS', 'Windows', 'Linux')")
      .filter("visit.visitedPage != ''")
      .select($"key", $"value")

    val writeQuery = recordsToWrite.writeStream
      .option("checkpointLocation", "/tmp/bde/module5/homework/e2/checkpoint")
      .format("kafka")
      .option("kafka.bootstrap.servers", "localhost:29092")
      .option("topic", "valid_visits")

    writeQuery.start().awaitTermination()
  }

}

File: 03-stream-processing/python/exercise2/streaming_job.py
from pyspark.sql import SparkSession

from pyspark.sql import functions as F
from pyspark.sql.types import StructType, StringType, StructField

spark = SparkSession.builder.master("local[*]") \
    .config('spark.jars.packages', 'org.apache.spark:spark-sql-kafka-0-10_2.12:3.4.0') \
    .getOrCreate()

input_data_stream = spark.readStream \
    .option("kafka.bootstrap.servers", "localhost:29092") \
    .option("subscribe", "visits_raw") \
    .option("startingOffsets", "EARLIEST") \
    .format("kafka") \
    .load()

user_visit_schema = StructType([
    StructField("os", StringType()),
    StructField("visitedPage", StringType())
])

records_to_write = input_data_stream.select(
        F.col("key").cast("string"),
        F.from_json(F.col("value").cast("string"), user_visit_schema).alias("visit"),
        F.col("value").cast("string")
    ).filter("visit.os IN ('macOS', 'Windows', 'Linux')").filter("visit.visitedPage != ''") \
        .select(
        F.col("key"), F.col("value")
    )

write_query = records_to_write.writeStream.format("kafka")\
    .option("kafka.bootstrap.servers", "localhost:29092").option("topic", "valid_visits")\
    .start()

write_query.awaitTermination()
